compiler/rustc_parse/src/parser/mod.rs

```diff
@@ -1260,9 +1260,11 @@ impl<'a> Parser<'a> {
         let start_token = (self.token.clone(), self.token_spacing);
         let cursor_snapshot = TokenCursor {
             frame: self.token_cursor.frame.clone(),
-            // We only ever capture tokens within our current frame,
-            // so we can just use an empty frame stack
-            stack: vec![],
+            // If we start collecting on the first token of a frame
+            // (e.g. the '{' in '{ my_tokens }'), then we can legally
+            // exit the frame (e.g. collecting '{ my_tokens } something_else'
+            // Therefore, we preserve the top of the stack, but nothing else
+            stack: self.token_cursor.stack.first().cloned().into_iter().collect(),
```

Comment:
```suggestion
            stack: if self.token_cursor.stack.is_empty() { vec![] } else { vec![TokenCursorFrame::dummy()] }
```
I'm not actually sure, does it matter what exactly we put into the stack here, or putting any dummy frame making `stack.pop()` work without panic would be enough?

I'm asking because `first()` is not the top of the stack, it's the bottom, but things still seem to work.

Reply:
Oh, that's a good point. I'll need to take a closer look at how this is working.